File: Assignment_4/Python/scheduler/scheduler.py
"""
Python/scheduler/scheduler.py
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Two-pass scheduling algorithm with preference-aware gap filling.

Pass 1 — Preferred Assignment
    For every (day, employee) pair (employees shuffled per day for
    fairness), assign the employee's highest-ranked shift that
    doesn't violate the 5-day cap or the one-shift-per-day rule.

Pass 2 — Gap Filling
    Scan every (day, shift) slot. If headcount < MIN_EMPLOYEES_PER_SHIFT,
    pull available employees and sort them by preference_rank for that
    shift — so the person who actually wants that slot goes first.
"""

import random
from models.employee import Employee, DAYS, SHIFTS, MIN_EMPLOYEES_PER_SHIFT


class Scheduler:
    """Builds the weekly shift schedule for a pool of employees."""

    def __init__(self, employees: list[Employee]) -> None:
        self.employees: list[Employee] = employees
        # schedule[day][shift] → [name, ...]
        self.schedule: dict[str, dict[str, list[str]]] = {
            day: {shift: [] for shift in SHIFTS} for day in DAYS
        }
        self.warnings: list[str] = []

    # ── Public entry ─────────────────────────────────────────────────────────
    def build(self) -> None:
        """Reset, then run both scheduling passes."""
        self._reset()
        self._pass1_preferred()
        self._pass2_fill_gaps()

    # ── Pass 1 ───────────────────────────────────────────────────────────────
    def _pass1_preferred(self) -> None:
        for day in DAYS:
            pool = self.employees[:]
            random.shuffle(pool)                    # fairness: no fixed order

            for emp in pool:
                if not emp.is_available():
                    continue
                if emp.is_working_on(day):
                    continue

                # Walk ranked preferences; first valid shift wins.
                # "Conflict" = employee already assigned or capped (handled
                # by the guards above). If all preferences blocked, the
                # employee stays free and may be pulled in Pass 2.
                for shift in emp.preferences:
                    self._assign(emp, day, shift)
                    break

    # ── Pass 2 ───────────────────────────────────────────────────────────────
    def _pass2_fill_gaps(self) -> None:
        for day in DAYS:
            for shift in SHIFTS:
                while len(self.schedule[day][shift]) < MIN_EMPLOYEES_PER_SHIFT:
                    available = [
                        e for e in self.employees
                        if not e.is_working_on(day) and e.is_available()
                    ]
                    if not available:
                        self.warnings.append(
                            f"Cannot fill '{shift}' on {day}: "
                            "not enough available employees."
                        )
                        break

                    # Bonus: sort by preference rank so the best-match person
                    # is always chosen first even during forced assignments.
                    available.sort(key=lambda e: e.preference_rank(shift))
                    self._assign(available[0], day, shift)

    # ── Helper ───────────────────────────────────────────────────────────────
    def _assign(self, emp: Employee, day: str, shift: str) -> None:
        emp.assign(day, shift)
        self.schedule[day][shift].append(emp.name)

    def _reset(self) -> None:
        for emp in self.employees:
            emp.reset()
        self.schedule  = {day: {shift: [] for shift in SHIFTS} for day in DAYS}
        self.warnings  = []

    # ── Reporting helpers (used by GUI and CLI) ───────────────────────────────
    def get_employee_summary(self) -> list[dict]:
        """
        Returns one dict per employee (sorted by name):
            { name, days_worked, preferences, days: [(day, shift), …] }
        """
        out = []
        for emp in sorted(self.employees, key=lambda e: e.name):
            out.append({
                "name":        emp.name,
                "days_worked": emp.days_worked,
                "preferences": emp.preferences,
                "days": [
                    (d, emp.schedule[d])
                    for d in DAYS
                    if emp.is_working_on(d)
                ],
            })
        return out

    # ── CLI text output ───────────────────────────────────────────────────────
    def print_schedule(self) -> None:
        col = 26
        header = f"{'Day':<13}" + "".join(
            f"{s.capitalize():<{col}}" for s in SHIFTS
        )
        border = "=" * len(header)
        sep    = "-" * len(header)

        print(border)
        print("         WEEKLY EMPLOYEE SCHEDULE")
        print(border)
        print(header)
        print(sep)
        for day in DAYS:
            row = f"{day:<13}"
            for shift in SHIFTS:
                names = self.schedule[day][shift]
                cell  = ", ".join(names) if names else "(empty)"
                row  += f"{cell:<{col}}"
            print(row)
        print(border)

        if self.warnings:
            print("\n[WARNINGS]")
            for w in self.warnings:
                print(f"  ⚠  {w}")

    def print_employee_summary(self) -> None:
        print("\n         EMPLOYEE SUMMARY")
        print("-" * 56)
        STARS = {0: "★★★", 1: "★★☆", 2: "★☆☆"}
        for row in self.get_employee_summary():
            print(f"  {row['name']:<24} Days worked: {row['days_worked']}")
            for day, shift in row["days"]:
                rank  = row["preferences"].index(shift)
                print(f"    {day:<13} → {shift:<11} {STARS[rank]}")
        print("-" * 56)

File: Assignment_4/Cpp/src/scheduler.cpp
/**
 * Cpp/src/scheduler.cpp
 * ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 * Scheduler implementation — two-pass algorithm.
 *
 * FIX (build error):
 *   getEmployeeSummary() now returns QVector<EmployeeSummaryRow>
 *   (a typed struct defined in scheduler.h) instead of
 *   QVector<QMap<QString,QVariant>>.  This eliminates all
 *   "invalid use of incomplete type 'class QVariant'" errors that
 *   appeared with Qt 6.10 + GCC 15 because QVariant was only
 *   forward-declared when pulled in through QMap's include chain.
 */

#include "scheduler.h"

#include <QTextStream>
#include <algorithm>
#include <numeric>

// ── Constructor ───────────────────────────────────────────────────────────────
Scheduler::Scheduler(QVector<Employee> &employees)
    : schedule(NUM_DAYS, QVector<QStringList>(NUM_SHIFTS)), employees_(employees)
{
}

// ── Public: build ─────────────────────────────────────────────────────────────
void Scheduler::build()
{
    reset();
    assignPreferredShifts();
    fillGaps();
}

// ── Pass 1: preferred assignment ──────────────────────────────────────────────
void Scheduler::assignPreferredShifts()
{
    for (int d = 0; d < NUM_DAYS; ++d)
    {
        // Shuffle employee indices each day for fairness (no fixed priority)
        QVector<int> idx(employees_.size());
        std::iota(idx.begin(), idx.end(), 0);

        for (int i = idx.size() - 1; i > 0; --i)
        {
            int j = static_cast<int>(
                QRandomGenerator::global()->bounded(static_cast<quint32>(i + 1)));
            std::swap(idx[i], idx[j]);
        }

        for (int k : idx)
        {
            Employee &emp = employees_[k];
            if (!emp.isAvailable())
                continue;
            if (emp.isWorkingOn(d))
                continue;

            // Walk the ranked preference list; assign the first valid shift.
            // Conflict resolution: if the employee is capped (isAvailable)
            // or already placed today (isWorkingOn), the guards skip them.
            // Any remaining unassigned employees are picked up in Pass 2.
            for (const QString &shift : emp.preferences)
            {
                doAssign(emp, d, shift);
                break; // assigned — move to next employee
            }
        }
    }
}

// ── Pass 2: fill gaps ─────────────────────────────────────────────────────────
void Scheduler::fillGaps()
{
    for (int d = 0; d < NUM_DAYS; ++d)
    {
        for (int s = 0; s < NUM_SHIFTS; ++s)
        {
            while (schedule[d][s].size() < MIN_EMPLOYEES_PER_SHIFT)
            {
                // Collect employees not yet placed today and still available
                QVector<int> available;
                for (int i = 0; i < employees_.size(); ++i)
                {
                    const Employee &e = employees_[i];
                    if (!e.isWorkingOn(d) && e.isAvailable())
                        available.append(i);
                }

                if (available.isEmpty())
                {
                    warnings << QString("Cannot fill '%1' on %2: "
                                        "not enough available employees.")
                                    .arg(SHIFTS[s])
                                    .arg(DAYS[d]);
                    break;
                }

                // Bonus: sort by preference rank so the best-matched
                // employee is always chosen first even in forced assignments.
                std::sort(available.begin(), available.end(),
                          [&](int a, int b)
                          {
                              return employees_[a].preferenceRank(SHIFTS[s]) <
                                     employees_[b].preferenceRank(SHIFTS[s]);
                          });

                doAssign(employees_[available[0]], d, SHIFTS[s]);
            }
        }
    }
}

// ── Helper: assign + record ───────────────────────────────────────────────────
void Scheduler::doAssign(Employee &emp, int dayIdx, const QString &shift)
{
    emp.assign(dayIdx, shift);
    int sIdx = indexOfStr(SHIFTS, shift);
    if (sIdx >= 0)
        schedule[dayIdx][sIdx] << emp.name;
}

// ── Reset ─────────────────────────────────────────────────────────────────────
void Scheduler::reset()
{
    for (auto &e : employees_)
        e.reset();
    schedule = QVector<QVector<QStringList>>(
        NUM_DAYS, QVector<QStringList>(NUM_SHIFTS));
    warnings.clear();
}

// ── Reporting: typed summary (NO QVariant) ────────────────────────────────────
QVector<EmployeeSummaryRow> Scheduler::getEmployeeSummary() const
{
    // Collect const pointers, then sort by name
    QVector<const Employee *> sorted;
    sorted.reserve(employees_.size());
    for (const auto &e : employees_)
        sorted << &e;

    std::sort(sorted.begin(), sorted.end(),
              [](const Employee *a, const Employee *b)
              { return a->name < b->name; });

    QVector<EmployeeSummaryRow> out;
    out.reserve(sorted.size());

    for (const Employee *ep : sorted)
    {
        EmployeeSummaryRow row;
        row.name = ep->name;
        row.daysWorked = ep->daysWorked;
        row.preferences = ep->preferences;

        // Build per-day map: { "Monday" -> "morning", … }
        for (int d = 0; d < NUM_DAYS; ++d)
            if (ep->isWorkingOn(d))
                row.dayMap[DAYS[d]] = ep->schedule[d];

        out << row;
    }
    return out;
}

// ── CLI: print weekly schedule ────────────────────────────────────────────────
void Scheduler::printSchedule() const
{
    QTextStream out(stdout);
    const int COL = 26;
    const int DCOL = 13;
    const int total = DCOL + COL * NUM_SHIFTS;

    const QString dbl(total, '=');
    const QString sep(total, '-');

    out << dbl << "\n";
    out << "         WEEKLY EMPLOYEE SCHEDULE\n";
    out << dbl << "\n";

    // Header row
    out << QString("Day").leftJustified(DCOL);
    for (const auto &s : SHIFTS)
        out << QString(s[0].toUpper() + s.mid(1)).leftJustified(COL);
    out << "\n"
        << sep << "\n";

    // Data rows
    for (int d = 0; d < NUM_DAYS; ++d)
    {
        out << DAYS[d].leftJustified(DCOL);
        for (int s = 0; s < NUM_SHIFTS; ++s)
        {
            QString cell = schedule[d][s].isEmpty()
                               ? "(empty)"
                               : schedule[d][s].join(", ");
            out << cell.leftJustified(COL);
        }
        out << "\n";
    }
    out << dbl << "\n";

    if (!warnings.isEmpty())
    {
        out << "\n[WARNINGS]\n";
        for (const auto &w : warnings)
            out << "  ⚠  " << w << "\n";
    }
    out.flush();
}

// ── CLI: print employee summary ───────────────────────────────────────────────
void Scheduler::printEmployeeSummary() const
{
    QTextStream out(stdout);
    const QString STARS[] = {"★★★", "★★☆", "★☆☆"};

    out << "\n         EMPLOYEE SUMMARY\n";
    out << QString(56, '-') << "\n";

    const auto summary = getEmployeeSummary();
    for (const auto &row : summary)
    {
        out << "  " << row.name.leftJustified(24)
            << "Days worked: " << row.daysWorked << "\n";

        for (const QString &day : DAYS)
        {
            if (row.dayMap.contains(day))
            {
                const QString &shift = row.dayMap[day];
                int rank = row.preferences.indexOf(shift);
                QString stars = (rank >= 0 && rank < 3) ? STARS[rank] : "?";
                out << "    " << day.leftJustified(13)
                    << " -> " << shift.leftJustified(12) << stars << "\n";
            }
        }
    }
    out << QString(56, '-') << "\n";
    out.flush();
}
